src/cdk/datetime/datepicker-input.ts

```diff
@@ -0,0 +1,324 @@
+
+  /** The form control validator for whether the input parses. */
+  private _parseValidator: ValidatorFn = (): ValidationErrors | null => {
+    return this._lastValueValid ?
+      null : {'datepickerParse': {'text': this._elementRef.nativeElement.value}};
+  }
+
+  /** The form control validator for the min date. */
+  private _minValidator: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return (!this.min || !controlValue ||
+      this._dateAdapter.compareDate(this.min, controlValue) <= 0) ?
+        null : {'datepickerMin': {'min': this.min, 'actual': controlValue}};
+  }
+
+  /** The form control validator for the max date. */
+  private _maxValidator: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return (!this.max || !controlValue ||
+      this._dateAdapter.compareDate(this.max, controlValue) >= 0) ?
+        null : {'datepickerMax': {'max': this.max, 'actual': controlValue}};
+  }
+
+  /** The form control validator for the date filter. */
+  private _filterValidator: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return !this._dateFilter || !controlValue || this._dateFilter(controlValue) ?
+      null : {'datepickerFilter': true};
+  }
+
+  /** The combined form control validator for this input. */
+  private _validator: ValidatorFn | null =
+    Validators.compose(
+      [this._parseValidator, this._minValidator, this._maxValidator, this._filterValidator]);
+
+  /** Whether the last value set on the input was valid. */
+  private _lastValueValid = false;
+
+  /** Constructor for datepicker input. */
+  constructor(
+    private _elementRef: ElementRef,
+    @Optional() public _dateAdapter: DateAdapter<D>) {
+    if (!this._dateAdapter) {
+      throw Error('CdkDatepicker: No provider found for DateAdapter.');
+    }
+
+    // Update the displayed date when the locale changes.
+    this._localeSubscription = _dateAdapter.localeChanges.subscribe(() => {
+      this.value = this.value;
+    });
+  }
+
+  /** Content initialization. */
+  ngAfterContentInit() {
+    if (this._datepicker) {
+      this._datepickerSubscription = this._datepicker._selectedChanged.subscribe((selected: D) => {
+        this.value = selected;
+        this._controlValueAccessorOnChange(selected);
+        this._onTouched();
+        this.dateInput.emit(new DatepickerInputEvent(this, this._elementRef.nativeElement));
+        this.dateChange.emit(new DatepickerInputEvent(this, this._elementRef.nativeElement));
+      });
+    }
+  }
+
+  /** Destroys subscriptions and change detection. */
+  ngOnDestroy() {
+    this._datepickerSubscription.unsubscribe();
```

Comment:
use an aggregate subscription that we add other subscriptions to so we can just do a single unsubscribe here